src/main/java/bio/terra/validation/StudyRequestValidator.java

```diff
@@ -0,0 +1,203 @@
+        List<ColumnModel> columns = table.getColumns();
+
+        if (tableName != null && columns != null) {
+            List<String> columnNames = columns.stream().map(ColumnModel::getName).collect(toList());
+            if (Utils.hasDuplicates(columnNames)) {
+                errors.rejectValue("schema", "DuplicateColumnNames");
+            }
+            context.addTable(tableName, columnNames);
+        }
+    }
+
+    private void validateRelationshipTerm(RelationshipTermModel term, Errors errors, SchemaValidationContext context) {
+        String table = term.getTable();
+        String column = term.getColumn();
+        if (table != null && column != null) {
+            if (context.isInvalidTableColumn(table, column)) {
+                errors.rejectValue("schema", "InvalidRelationshipTermTableColumn");
+            }
+        }
+    }
+
+    private void validateRelationship(RelationshipModel relationship, Errors errors, SchemaValidationContext context) {
+        RelationshipTermModel fromTerm = relationship.getFrom();
+        if (fromTerm != null) {
+            validateRelationshipTerm(fromTerm, errors, context);
+        }
+
+        RelationshipTermModel toTerm = relationship.getTo();
+        if (toTerm != null) {
+            validateRelationshipTerm(toTerm, errors, context);
+        }
+
+        String relationshipName = relationship.getName();
+        if (relationshipName != null) {
+            context.addRelationship(relationshipName);
+        }
+    }
+
+    private void validateAssetTable(AssetTableModel assetTable, Errors errors, SchemaValidationContext context) {
+        String tableName = assetTable.getName();
+        List<String> columnNames = assetTable.getColumns();
+        if (tableName != null && columnNames != null) {
+            // An empty list acts like a wildcard to include all columns from a table in the asset specification.
+            if (columnNames.size() == 0) {
+                if (context.isInvalidTable(tableName)) {
+                    errors.rejectValue("schema", "InvalidAssetTable");
+                }
+            } else {
+                boolean anyInvalidTableColumns = columnNames.stream()
+                        .anyMatch((columnName) -> context.isInvalidTableColumn(tableName, columnName));
+
+                if (anyInvalidTableColumns) {
+                    errors.rejectValue("schema", "InvalidAssetTableColumn");
+                }
+            }
+        }
+    }
+
+    private void validateAsset(AssetModel asset, Errors errors, SchemaValidationContext context) {
+        List<AssetTableModel> assetTables = asset.getTables();
+        if (assetTables != null) {
+            if (assetTables.stream().noneMatch(AssetTableModel::isIsRoot)) {
+                errors.rejectValue("schema", "NoRootTable");
+            }
+            assetTables.forEach((assetTable) -> validateAssetTable(assetTable, errors, context));
+        }
+        List<String> follows = asset.getFollow();
```

Comment:
Is it worth checking connectivity of the graph here? My thought is no, it is complex and maybe not something we want in the long-term. But wanted to ask. Maybe comment as to why not?